### PR TITLE
perf(duckdb): use builtin negative array slicing in newer duckdbs

### PR DESCRIPTION
**Pull request by NickCrews**
Similar rationale to https://github.com/ibis-project/ibis/pull/10410

### CODE REVIEW

File: ibis/backends/sql/compilers/duckdb.py [hunk 1]
```diff
@@ -177,16 +177,19 @@ def visit_ArrayRepeat(self, op, *, arg, times):
         return self.f.flatten(self.f.list_apply(self.f.range(times), func))
 
     def visit_ArraySlice(self, op, *, arg, start, stop):
+        import duckdb
```
Comment (cpcloud) [inline ibis/backends/sql/compilers/duckdb.py:180]: The compilers **must** be usable without any dependencies installed, so you can't do this.

File: ibis/backends/sql/compilers/duckdb.py [hunk 2]
```diff
@@ -177,16 +177,19 @@ def visit_ArrayRepeat(self, op, *, arg, times):
...
             start = 0
-        else:
-            start = self.f.least(arg_length, self._neg_idx_to_pos(arg, start))
-
         if stop is None:
             stop = arg_length
-        else:
+
+        # TODO: once we only support duckdb >= 0.9.0, we can remove this
+        if duckdb.__version__ < "0.9.0":
```
Comment (cpcloud) [inline ibis/backends/sql/compilers/duckdb.py:190]: We should just bump to 0.9.0 as a lower bound for the next release, then we don't have to futz with with this.

I'll update your PR to make these changes.

**Review by cpcloud — CHANGES_REQUESTED** [review]

**Review by cpcloud — APPROVED** [review]

**Comment by cpcloud** [comment]
Well, clearly negative slicing doesn't work as expected 😅

**Comment by cpcloud** [comment]
This should replace things like `-3 + -1` with the literal `-4`.

File: ibis/backends/sql/compilers/duckdb.py (hunk 1 above)
Comment (NickCrews) [reply ibis/backends/sql/compilers/duckdb.py]: Ah, didn't know this. Thanks!

File: ibis/backends/sql/compilers/duckdb.py [hunk 3]
```diff
@@ -182,16 +182,22 @@ def visit_ArraySlice(self, op, *, arg, start, stop):
...
 
         if start is None:
-            start = 0
+            start = 1
+        elif isinstance(op.start, ops.Literal):
+            start = op.start.value
+            start += start >= 0
         else:
-            start = self.f.least(arg_length, self._neg_idx_to_pos(arg, start))
+            start = self.if_(start < 0, start, start + 1)
```
Comment (NickCrews) [inline ibis/backends/sql/compilers/duckdb.py:190]: Duckdb why do you have to use such inconsistent indexing semantics 😢

**Comment by NickCrews** [comment]
@cpcloud did you bump the min version to 0.10.0 because I didn't read duckdbs changelog/got blame correctly, and 0.9.0 wasn't adequate?

**Comment by cpcloud** [comment]
> @cpcloud did you bump the min version to 0.10.0 because I didn't read duckdbs changelog/got blame correctly, and 0.9.0 wasn't adequate?

0.10 is the first release that officially supports the backwards compatible 1.0 storage format, and it didn't seem like bumping to just 0.9 was worth it given that bumping to either is a breaking change from the previous lower bound, so I bumped to 0.10.

**Comment by NickCrews** [comment]
Sounds reasonable! If someone squawks that they need 0.9.0 support we can revisit

**Pull request merged**